src/governance.cpp

```diff
@@ -918,14 +919,18 @@ bool CGovernanceManager::MasternodeRateCheck(const CGovernanceObject& govobj, up
 
     switch(eUpdateLast) {
     case UPDATE_TRUE:
+        // we should update the buffer only if object has accepted
+        if (!fRateOK)
+            LogPrintf("CGovernanceManager::MasternodeRateCheck -- WARNING !!! Rate too high but eUpdateLast = UPDATE_TRUE : object hash = %s, masternode vin = %s, object timestamp = %d, rate = %f, max rate = %f\n",
+                        strHash, vin.prevout.ToStringShort(), nTimestamp, dRate, dMaxRate);
         pBuffer->AddTimestamp(nTimestamp);
         it->second.fStatusOK = fRateOK;
-        break;
+        return true;
     case UPDATE_FAIL_ONLY:
-        if(!fRateOK) {
-            pBuffer->AddTimestamp(nTimestamp);
+        if(!fRateOK)
```

Comment:
It looks like the problems you discovered still aren't fixed here.

The break is still missing for the UPDATE_FAIL_ONLY case and all code paths still return true.

Also I don't think UPDATE_TRUE should return true is the rate check fails.